build: add non root user for dynamo backend runtime containers

nv-anants opened the pull request:
#### Overview:

Transitioning all Dynamo backend containers to run as a dedicated non-root user instead of the default root user. 

Changes: 
* Create dedicated dynamo user (UID 1000, GID 0) across all framework containers
* GID 0 (root group) for openshift pattern - https://www.redhat.com/en/blog/a-guide-to-openshift-and-uids
* All runtime artifacts (wheels, venv..) and workspace directories owned by dynamo user
* Add a framework stage in TRTLLM image to allow installation as root and copy venv out of framework image into runtime image
* dev images will still use root user as default
* Update local_dev image to match dynamo user
* Removed host workspace mounts from test workflows
* Added --user flag support to container run with root user


Closes: OPS-1181


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added --user flag to the container launcher for running containers as a specified user.

* **Improvements**
  * Containers now default to a dedicated non-root runtime user and preconfigured Python virtualenv.
  * Reorganized build stages and base images for better compatibility and runtime packaging.
  * File ownership and workspace paths made consistent for predictable dev/runtime behavior.

* **Bug Fixes / CI**
  * Test runner now uses a stable test-results path, ensures results are created and copied out; workflow removed host workspace mounts.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nv-anants committed:
build: add non root user for dynamo dockerfile

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
update dockerfile.vllm to use dynamo user

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
add non root user to dockerfile trtllm

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
remove sudo

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
add missing stuff

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
add non root user for sglang

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
local dev and build updates

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
precommit

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
add local evn path

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
update ids and add sudo in dev

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
fix test permissions

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
fix pytest

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
remove group

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants commented:
@coderabbitai full review

nv-anants committed:
fixes

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
mkdir fix

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
pytest cp

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
fix pytest

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
remove ws mount

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
local dev updates

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
minor

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
temp: debug

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
fix

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
Revert "fix"

This reverts commit c37d61f396a2104f523fc29a47d648018ecb6895.

nv-anants committed:
Revert "temp: debug"

This reverts commit dc0f93d021d66225dcacccd050da6b15919298e8.

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
fxi

Signed-off-by: Anant Sharma <anants@nvidia.com>

keivenchang reviewed (commented):
I see a lot of `chown -R ...`. Do you know if the build time increased?

Also, can you update .devcontainer/*.j2, from ubuntu to dynamo user?

nv-anants commented on container/Dockerfile.vllm:
```diff
@@ -222,15 +222,28 @@ COPY --from=dynamo_base /usr/local/bin/etcd/ /usr/local/bin/etcd/
+    && useradd -m -s /bin/bash -g 0 dynamo \
+    && [ `id -u dynamo` -eq 1000 ] \
+    && mkdir -p /home/dynamo/.cache /opt/dynamo \
+    && chown -R dynamo: /workspace /home/dynamo /opt/dynamo \
```
yes, i have seen that as well. At this point in time, the directories are (almost) empty so its much faster to do before we start adding stuff

nnshah1 commented on container/Dockerfile.sglang:
```diff
@@ -450,8 +464,13 @@ RUN apt-get update -y && \
     cloc \
     locales \
+    # sudo for dev stage
+    sudo \
```
we need sudo for installing packages, etc. is that right?

nv-anants replied:
yes, so dev stage will have sudo permissions but not runtime (as not expecting anyone to change runtime contianer)

nv-anants commented on container/Dockerfile.trtllm:
trtllm installation need admin permission to update and overwrite some files - https://github.com/ai-dynamo/dynamo/blob/anants/no-root-user/container/Dockerfile.trtllm#L126-L159

So i have split out all the trtllm stuff in a multi stage where it has admin permissions and then after install i copy out venv in runtime container with non root user ownership on venv files

nv-anants commented on container/Dockerfile.vllm:
```diff
@@ -222,15 +222,28 @@ COPY --from=dynamo_base /usr/local/bin/etcd/ /usr/local/bin/etcd/
+ENV HOME=/home/dynamo
+
+# Copy UCX and NIXL to system directories
+COPY --chown=dynamo: --from=dynamo_base /usr/local/ucx /usr/local/ucx
```
either way should be fine as we will only read these files i assume. But we are building it in dynamo dockerfile, so i kept it as dynamo owned

nv-tusharma commented on container/Dockerfile.vllm:
```diff
@@ -273,16 +284,16 @@ RUN --mount=type=bind,source=./container/deps/requirements.txt,target=/tmp/requi
 RUN --mount=type=bind,source=./container/launch_message.txt,target=/workspace/launch_message.txt \
-    sed '/^#\s/d' /workspace/launch_message.txt > ~/.launch_screen && \
-    echo "cat ~/.launch_screen" >> ~/.bashrc && \
-    echo "source $VIRTUAL_ENV/bin/activate" >> ~/.bashrc
```
Any reason to remove this? If I do /bin/bash into the container, will I still default into the python virtual environment?

nv-tusharma commented on container/Dockerfile.sglang:
```diff
@@ -467,12 +486,14 @@ RUN curl --retry 3 --retry-delay 2 -LSso /usr/local/bin/clang-format https://git
-COPY pyproject.toml README.md hatch_build.py /workspace/
-RUN pip install --no-deps -e .
+COPY --chown=dynamo: pyproject.toml README.md hatch_build.py /workspace/
+RUN python3 -m pip install --no-deps -e .
```
Curious, does pip install not work anymore? It needs the python3 -m prefix?

nv-tusharma commented on container/run.sh:
```diff
@@ -183,6 +184,14 @@ get_options() {
                 missing_requirement "$1"
             fi
             ;;
+        --user)
```
If we're adding sudo permissions to dynamo user, in dev stage, is there any reason for why we need to manually specify the --user in run.sh?

nv-anants committed:
update to dynamo

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nnshah1 reviewed (approved):
looks good to me, a few things to check:

1) can we intialize the environment for any user? (otherwise it won't work for non 100 user)
2) when volume mounting you should also pass in the users UID and GID 

In the past the pattern I had done in a run.sh like thing -> if there are any volume mounts then default the USER and GID passed into the container (UID and GID - names won't work) - so that then the access to the volume mounts would work.

 However since the environment is set up for dynamo - I'm not sure that will work here

The way that vscode / devcontainer does it is to inject a change to rewrite the UID and GID of the remote user to the local user - not sure that can be automated in run.sh as well -

nnshah1 commented on container/Dockerfile:
```diff
@@ -380,24 +379,40 @@ RUN --mount=type=bind,source=./container/deps/requirements.txt,target=/tmp/requi
-    sed '/^#\s/d' /opt/dynamo/launch_message.txt > ~/.launch_screen && \
-    echo "cat ~/.launch_screen" >> ~/.bashrc && \
-    echo "source $VIRTUAL_ENV/bin/activate" >> ~/.bashrc
+    sed '/^#\s/d' /opt/dynamo/launch_message.txt > ${HOME}/.launch_screen && \
```
can you test with passing in a different user?

nnshah1 commented on container/run.sh:
```diff
@@ -183,6 +184,14 @@ get_options() {
                 missing_requirement "$1"
             fi
             ;;
+        --user)
```
but I thought there was also a way to use --group separately

nv-anants committed:
Merge branch 'main' into anants/no-root-user

nv-anants committed:
review fixes

1. write init commands to /etc/bash.bashrc
2. set root as default user in dev container
3. update --user help message

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
ubuntu -> dynamo

Signed-off-by: Anant Sharma <anants@nvidia.com>

nv-anants committed:
Merge branch 'main' into anants/no-root-user
